pkg/util/git.go

```diff
@@ -0,0 +1,203 @@
+)
+
+var (
+	KubernetesGitHubURL     = fmt.Sprintf("https://github.com/%s/kubernetes", githubOrg)
+	KubernetesGitHubAuthURL = fmt.Sprintf("%s%s/kubernetes.git", githubAuthRoot, githubOrg)
+)
+
+func BranchExists(branch string) (bool, error) {
+	log.Printf("Verifying %s branch exists on the remote...", branch)
+
+	// Create the remote with repository URL
+	rem := git.NewRemote(
+		memory.NewStorage(),
+		&config.RemoteConfig{
+			Name: "origin",
+			URLs: []string{KubernetesGitHubURL},
+		},
+	)
+
+	log.Print("Fetching tags...")
+
+	// We can then use every Remote functions to retrieve wanted information
+	refs, err := rem.List(&git.ListOptions{})
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	// Filters the references list and only keeps branches
+	for _, ref := range refs {
+		if ref.Name().IsBranch() {
+			if ref.Name().Short() == branch {
+				log.Printf("Found branch %s", ref.Name().Short())
+				return true, nil
+			}
+		}
+	}
+
+	log.Printf("Could not find branch %s", branch)
+
+	return false, nil
+}
+
+// TODO: Need to handle https and ssh auth sanely
+func SyncRepo(repoURL, destination string) error {
+	log.Printf("Syncing %s to %s:", repoURL, destination)
+
+	if _, err := os.Stat(destination); !os.IsNotExist(err) {
+		chdirErr := os.Chdir(destination)
+		if chdirErr != nil {
+			return chdirErr
+		}
+
+		repo, repoErr := git.PlainOpen(destination)
+		if repoErr != nil {
+			return repoErr
+		}
+
+		w, err := repo.Worktree()
+		//CheckIfError(err)
+		if err != nil {
+			return err
+		}
+
+		// ... checking out to commit
+		//Info("git checkout %s", commit)
+		err = w.Checkout(&git.CheckoutOptions{
+			Branch: plumbing.NewBranchReferenceName("master"),
+		})
+		//CheckIfError(err)
+		if err != nil {
+			return err
+		}
+
+		pullOpts := &git.PullOptions{RemoteName: "origin"}
+		err = w.Pull(pullOpts)
+		if err != nil && err != git.NoErrAlreadyUpToDate {
+			return err
+		}
+
+		return nil
+	}
+
+	cloneOpts := &git.CloneOptions{
+		URL: repoURL,
+	}
+	_, repoErr := git.PlainClone(destination, false, cloneOpts)
+	if repoErr != nil {
+		return repoErr
+	}
+
+	return nil
+}
+
+// RevParse parses a git revision and returns a SHA1 on success, otherwise an
+// error.
+func RevParse(rev, workDir string) (string, error) {
+	repo, err := git.PlainOpen(workDir)
+	if err != nil {
+		return "", err
+	}
+
+	ref, err := repo.ResolveRevision(plumbing.Revision(rev))
+	if err != nil {
+		return "", err
+	}
+
+	return ref.String(), nil
+}
+
+// RevParseShort parses a git revision and returns a SHA1 on success, otherwise an
+// error.
+func RevParseShort(rev, workDir string) (string, error) {
+	fullRev, err := RevParse(rev, workDir)
+	if err != nil {
+		return "", err
+	}
+
+	shortRev := fullRev[:10]
+
+	return shortRev, nil
+}
+
+func GetMergeBase(masterRefShort, releaseRefShort string, repo *git.Repository) (string, error) {
+	masterRef := fmt.Sprintf("origin/%s", masterRefShort)
+	releaseRef := fmt.Sprintf("origin/%s", releaseRefShort)
```

Comment:
The default remote `origin` may be a `const` value somewhere since it's used in many places.

Reply:
Fixed!